### DISCUSSION TITLE
Understanding land sequence while manually flying

### DISCUSSION THREAD

**Discussion by puliceworx** [bitcraze/discussions#875] [Q&A] [answered]
Greetings, 

Scenario: I took a log of assisted takeoff and landed with a controller.  

It was fairly straightforward to figure out assisted takeoff and the hover setpoint.  
```
INFO:cflib.crazyflie.commander:send_hover_setpoint vx 0.000000 vy -0.000000 yr 0.000000 z 0.400000
INFO:cflib.crazyflie.commander:send_hover_setpoint vx 0.000000 vy -0.000000 yr 0.000000 z 0.400000
```
Repeated over and over from reading no change in the controller. 

However, this landing I can't seem to figure out what the client is doing to actually formulate these commands. 
It clearly has to be there somewhere but I can't seem to find it.  

It seems to me from the log that when the button is released this gets sent
INFO:cflib.crazyflie.commander:send_zdistance roll 0.000000 pitch 0.000000 yr 0.000000 z 0.400000
Is my assumption correct that the release of the assist button triggers this packet to be sent? 

Then a series of set points with decreased thrust until on the ground.
```
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 28899.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 26910.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 26109.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 25194.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 23966.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 23244.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 22287.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 21918.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 21624.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 21392.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 20670.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 20435.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 20201.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 19890.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 19030.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 18778.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 18485.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 18251.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 17784.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 17510.000000
INFO:cflib.crazyflie.commander:send_setpoint roll 0.000000 pitch 0.000000 yr 0.000000 thrust 16846.000000

```
This leaves me with two questions really:
1) Where is the loop that creates this sequence of setpoints? 
2) Why do I not see a Zero thrust setpoint to turn the motors off?  

I hope this makes sense and appreciate any pointers someone may have.

**Comment by krichardsson** [comment, marked as answer]
This code is fairly pretty "creative" and not that easy to digest... (meaning it should be rewritten)

> Where is the loop that creates this sequence of setpoints?

The source of the call is in [the joystick handling](https://github.com/bitcraze/crazyflie-clients-python/blob/d387b70d3f96be8c72f749ec4de5722f582bc2c5/src/cfclient/utils/input/__init__.py#L527-L529) where there is a call to `self.input_updated.call(...)`
The `input_updated` callback is [mapped to `send_setpoint()` in main.py](https://github.com/bitcraze/crazyflie-clients-python/blob/d387b70d3f96be8c72f749ec4de5722f582bc2c5/src/cfclient/ui/main.py#L196-L198) and this is what you see in your log

I think the client switches to non assisted mode and there is some sort of functionality for reducing thrust "softly" that kicks in.

The actual reduction in thrust is done in [inputreaderinterface.py](https://github.com/bitcraze/crazyflie-clients-python/blob/d387b70d3f96be8c72f749ec4de5722f582bc2c5/src/cfclient/utils/input/inputreaderinterface.py#L210-L212) where you can find 

```python 
                            lowering = ((current_time - self._last_time) *
                                        self.input.thrust_slew_rate)
                            limited_thrust = self._prev_thrust - lowering
```

> Why do I not see a Zero thrust setpoint to turn the motors off?

Not sure how you log this but a "zero" setpoint is sent all the time if there is no user input. I added a print in `send_setpoint()` and I can see the sequence with thrust going down and then an infinite flow of "zero" setpoints.

**Comment by puliceworx** [comment]
Wow lol  Ok I missed that part in the inputreader.  Thank you for pointing that out! 

I logged from cflib by adding a statement in each of the different setpoint sends.  
After rerunning I see the null setpoints.  I must have deleted the zero setpoints when I saved my log file.
